From ba6e35998b8d31804eadee3eb1a694e9de64536d Mon Sep 17 00:00:00 2001
From: Sidhanta Sahu <quic_sidhanta@quicinc.com>
Date: Thu, 6 Mar 2025 00:45:12 -0800
Subject: [PATCH 2/3] UPSTREAM:CP : wifi: ath12k: Support to manage primary
 peer lists

Add a peer item into the primary peers list when a station moves from
AUTH to ASSOC and remove it from the primary peers list when a station
moves from ASSOC to AUTH.

Add support to remove peer obects from the primary peers list during
recovery (mode-0 and mode-1).

Extend the RCU lock to manage the ath12k_peers in the primary peers list
and free the peer object using the call_rcu routine. The ath12k_peer
object has a member rcu_head, and the RCU read lock is used across the
critical path to access the peer object.

Patch-dependency: 858-wifi-ath12k-Add-debugfs-support-to-configure-primary.patch
                  970-QSDK-ath12k-Reuse-vp-profiles-in-case-of-splitphy.patch
Patch-work: None

Signed-off-by: Sidhanta Sahu <quic_sidhanta@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c |  1 +
 drivers/net/wireless/ath/ath12k/dp.c   |  1 +
 drivers/net/wireless/ath/ath12k/mac.c  | 28 ++++++++++++++++++++++++--
 3 files changed, 28 insertions(+), 2 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -3514,6 +3514,7 @@ static void ath12k_core_reset(struct wor
 		case ATH12K_MLO_RECOVERY_MODE0:
 			fallthrough;
 		default:
+			ath12k_peer_purge_primary_list();
 			ag->recovery_mode = ATH12K_MLO_RECOVERY_MODE0;
 		}
 		ath12k_dbg(ab, ATH12K_DBG_MODE1_RECOVERY,
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -130,6 +130,7 @@ int ath12k_dp_peer_setup(struct ath12k *
 
 	peer->primary_link = true;
 	arvif->primary_sta_link = true;
+	ath12k_peer_add_primary(peer);
 
 	if (vif->type == NL80211_IFTYPE_STATION)
 		ath12k_dp_tx_ppeds_cfg_astidx_cache_mapping(ar->ab, arvif, true);
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1413,6 +1413,10 @@ int ath12k_mac_partner_peer_cleanup(stru
 
 	ag = ab->ag;
 
+	spin_lock_bh(&ab->base_lock);
+	ath12k_peer_purge_primary_list_per_ab(ab);
+	spin_unlock_bh(&ab->base_lock);
+
 	for (idx = 0; idx < ag->num_chip; idx++) {
 		partner_ab = ag->ab[idx];
 
@@ -1514,7 +1518,7 @@ void ath12k_mac_peer_cleanup_all(struct
 	if (peer->sawf_ctx_peer.telemetry_peer_ctx)
 		ath12k_telemetry_peer_ctx_free(peer->sawf_ctx_peer.telemetry_peer_ctx);
 #endif
-		kfree(peer);
+		ath12k_peer_free(peer);
 	}
 	spin_unlock_bh(&ab->base_lock);
 	mutex_unlock(&ab->tbl_mtx_lock);
@@ -7674,6 +7678,8 @@ static int ath12k_station_disassoc(struc
 	struct ath12k_sta *ahsta = arsta->ahsta;
 	struct ieee80211_link_sta *link_sta;
 	struct ieee80211_sta *sta;
+	struct ath12k_peer *peer;
+	struct ath12k_base *ab;
 	int ret;
 
 	lockdep_assert_held(&ar->conf_mutex);
@@ -7706,6 +7712,22 @@ static int ath12k_station_disassoc(struc
 
 	ath12k_11s_mlo_testing_notify_peer_disassoc_h_mlo(arvif->ahvif, arsta);
 
+	ab = ar->ab;
+
+	spin_lock_bh(&ab->base_lock);
+
+	peer = ath12k_peer_find(ab, arsta->arvif->vdev_id, arsta->addr);
+	if (peer && peer->primary_link) {
+		ath12k_info(ab, "Disassoc current primary peer %pM peer_id 0x%x ml_peer_id 0x%x link_id 0x%x\n",
+			    peer->addr, peer->peer_id, peer->ml_peer_id, peer->link_id);
+
+		ath12k_peer_remove_primary(peer);
+		peer->primary_link = false;
+		arsta->arvif->primary_sta_link = false;
+	}
+
+	spin_unlock_bh(&ab->base_lock);
+
 	if (link_sta)
 		ath12k_mac_vendor_send_disassoc_event(arsta, link_sta);
 	if (!sta->wme) {
@@ -8249,7 +8271,7 @@ static void ath12k_mac_station_post_remo
 		if (peer->sawf_ctx_peer.telemetry_peer_ctx)
 			ath12k_telemetry_peer_ctx_free(peer->sawf_ctx_peer.telemetry_peer_ctx);
 #endif
-		kfree(peer);
+		ath12k_peer_free(peer);
 		ar->num_peers--;
 	}
 	spin_unlock_bh(&ar->ab->base_lock);
